working/digit-separators/feature-specification.md

```diff
@@ -0,0 +1,199 @@
+clear delimiters that do not occur anywher else), so this is unlikely to happen
+in practice. Underscore is already used by a large number of programming
+languages including Java, Swift, and Python.
+
+We also want to allow multiple separators for higher-level grouping, e.g.,:
+
+```none
+100__000_000_000__000_000_000
+```
+
+For this purpose, the underscore extends gracefully. So does space, but has the
+disadvantage that it collapses when inserted into HTML, whereas `''` looks odd.
+
+### Related work
+
+* [Java digit separators](https://docs.oracle.com/javase/8/docs/technotes/guides/language/underscores-literals.html)
+* [Python PEP 515 - underscores in numeric literals](https://peps.python.org/pep-0515/)
+
+### Possible new lint rules
+
+There are some possible new lint rule considerations, but none of these are
+considered vital to the usability or general success of the feature.
+
+The feature is designed to help the readability of long numbers. But a
+developer can still make an error about where to place separators. For example:
+
+```
+var one = 1_000_000;
+var two = 2_000_000;
+var three = 3_000_000;
+var four = 4_0000_000;
+```
+
+If a developer uses the Dart formatter to format their code, they cannot try to
+vertically align the numbers with whitespace (extra space characters are
+removed by the formatter). So we could offer a lint rule to only place
+separators every three digits of a decimal number. Also possibly a similar rule
+for hexadecimal numbers. If a developer ever uses digit separators for a
+different purpose (as in separating the digits of a phone number), the rule may
+not prove useful.
+
+A separate lint rule could encourage _consistent_ digit separators, which
+triggers if the digit groups do not have the same size (except the most
+significant I one, which can be shorter). If there are any `__` separators, the
+number of `_`-separated groups between them should also be the same, and
+repeatedly for higher numbers of `_`s.
+
+### Possible new quick fixes
+
+There are some possible new automated fix ("quick fix") considerations, but
+none of these are considered vital to the usability or general success of the
+feature.
+
+#### Unexpected underscores
+
+With the digit-separators feature, separators can be added between _digits_ of
+a number literal, but nowhere else. In most error cases, the unexpected
+underscore can be detected as such, and we can offer quick fixes to remove
+unexpected errors (for example, `100_`, `100_e.2`, `100._00`). In a few cases,
+the intention is not as straightforward, such as `100._100`, where `_100` can
+be a legal name of an extension member (though the presense of such a private
+extension member can be detected).
+
+#### Unexpected commas
+
+The only legal digit separator that is introduced with this feature is the
+underscore character. If a developer attempts to use another character, for
+example commas, as a separator, we may be able to detect this, and offer a
```

Comment:
Wouldn't this fall into the same problem that we have when we consider `,` as a separating digit? How can we be sure that the user intended to use `,` as a separating digit instead of two expressions separated by a comma?

Reply:
We can't. Many automated fixes are offered with the understanding that they are not the only fix, and may not be the desired fix.

I personally doubt we will offer this fix. But there might be some cases, like if the user pastes 100 lines of:

```dart
var x1 = 1,000,000;
var x2 = 2,000,000;
var x3 = 3,000,000;
```

into their editor, from another source text, we _could_ offer a fix to convert the commas, because it looks more likely that they meant to write `x1 = 1_000_000`, since  `var x1 = 1,` cannot be followed by an int literal (it can only be followed by an identifier.

Reply:
If the same person pasted 100 lines of `var x = foo(1,000,000);` we'd be having a harder time helping them.
(Not hat we couldn't *suggest* `1_000_000`, especially if we can see that `foo` takes only one argument, but it was not grammatically wrong to begin with.)

One thing that worries me is whether we can do the same fixes *after* the formatter has run.
If not, we may have a problem helping people who auto-format often.
If the code doesn't parse, it likely won't format either, but if it parses, and you get to `var x1 = foo(1, 000, 000);`, it would be sad if it's too late to help the user.